src/adapter/src/optimize/peek.rs

```diff
@@ -436,7 +496,8 @@ impl<'s> Optimize<GlobalMirPlan<ResolvedGlobal<'s>>> for Optimizer {
 
                 peek_plan
             }
-            None => {
+            // Should be slow path for copy to
+            (_, Some(_)) | (None, _) => {
                 // Ensure all expressions are normalized before finalizing.
                 for build in df_desc.objects_to_build.iter_mut() {
                     normalize_lets(&mut build.plan.0)?
```

Comment:
Since a copy-to-S3 is quite different from a peek (no fast path, no index, but a sink). I would expect to see this in a separate optimizer pipeline, e.g. `optimize/copy_to_s3.rs`. Is there a reason that wouldn't work?

Reply:
That's what I had originally, but it involved a lot of code duplication https://github.com/MaterializeInc/materialize/pull/24562#issuecomment-1915825181. COPY TO is a bit different from peek but it also shares a lot of same logic of spinning up a similar dataflow but with a sink in this case.

Reply:
Having looked at the changes  in `peek::Optimizer`, I have to say I strongly agree with @teskje here. Even if we use the same code at the `Coordinator` level the `Optimizer` level the difference in the optimizer pipeline is sufficient to warrant a different `optimize` module.

I'll try to pick up this branch and come up with something that satisfies both @MaterializeInc/compute and @MaterializeInc/surfaces as a proposal.